### DISCUSSION TITLE
Checksum of file while downloading

### DISCUSSION THREAD

**Discussion by etherx-dev** [curl/curl#6851] [Ideas]
mothballed due to lack of interest### 

_Originally posted by @jay in https://github.com/curl/curl/issues/1399#issuecomment-300685532_

**Comment by jay** [comment]
Unfortunately real time calculation of the sha256 checksum didn't have any interest.

**Comment by nmoinvaz** [comment]
There are some CDNs that will return response headers that contain a hash of the content but it is MD5 not SHA256. https://tools.ietf.org/html/rfc1864

**Comment by etherx-dev** [comment]
> There are some CDNs that will return response headers that contain a hash of the content but it is MD5 not SHA256. https://tools.ietf.org/html/rfc1864

See

> Unfortunately real time calculation of the sha256 checksum didn't have any interest.

Is there downside to this proposal as I cannot understand why this has been historically the explanation....

I mean how does this https://github.com/search?q=curl+%7C+sha+sum&type=Code  not show that there is interest.


Sorry for being harsh. I just need to understand if there is any other reason than “lack of interest”

**Comment by bagder** [comment]
> if there is any other reason than “lack of interest”

I don't think so. What would that be? There was no particular feedback, cheers or otherwise on @jay's work back then so I think that qualifies pretty good as "lack of interest".

**Comment by riptl** [comment]
I've seen a few software vendors supply hashes and signatures for downloads.
This is a common pattern for checking the signature of binary executables that are being installed outside of the reign of a package manager.

For example: https://releases.hashicorp.com/terraform/0.14.9/
- The `.zip` files are the actual download
- A `SHA256SUMS` file
- A GPG signature of the `SHA256SUMS` file

As op showed in #1399, it's common in Dockerfile build scripts that bundle out-of-band software.
However binaries are usually so small that there is no noticeable time save at all,
compared to the time it takes to download GPG keys and verify the signature.

Still, I would expect Dockerfile build scripts etc. to switch to this new feature if it gets added to reduce code complexity.

Another use-case is verifying large files (media, backups).
The average hard drive can read at 100 MB/s or so.
Hashing while downloading could save hours.

But I can think of a lot of design considerations.

1. Do we link libraries or compile them in statically?
2. Do we want to support CPU-specific instructions to speed up hashing? (probably 2-4x faster, but goes horribly wrong with illegal instruction exceptions if done wrong)
3. Which hashes do we support? (MD5, RIPEMD-160, SHA-1, SHA-2, SHA-3, Blake2, ...)
4. If we use a library, is it portable enough to work on all of curl's crazy architectures?
5. Should curl just output a hash?
6. Should curl verify against HTTP response headers or trailers that indicate hashes? (RFC 1544, RFC 3230)
6. Should curl be able to verify against a SUMS file?
7. If using a SUMS file, use sha256sum style or OpenSSL dgst style or both?
8. Should curl be able to verify against a remote SUMS file by URL?

**Comment by jayaddison** [comment]
Potentially relevant: the W3C SubResource Integrity (SRI) specification defines a string format for an `integrity` attribute, and associated semantics (roughly: consider only the strongest algorithm, but if there are multiple values with the same algorithm, permit any of them as valid), to be used validating fetched content.  Three algorithms (`sha256`, `sha384`, `sha512`) are [supported based on the CSP2 spec](https://www.w3.org/TR/CSP2/#hash_algo) that is referenced as a dependency.

Clearly it's possible to do that in a web browser where the source/DOM of the current page contains the expectations -- in the case of `curl` we have no 'current page' so we'd need to configure each request with content-integrity expectations.

There are probably good arguments _not_ to use the SRI `integrity` attribute format in `curl` -- too tightly coupling the technical dependencies, for example -- but I figured I'd mention it.
